File: 0090-subsets-ii/0090-subsets-ii.java
class Solution {
    HashSet<String> set = new HashSet<>();
    public List<List<Integer>> subsetsWithDup(int[] nums) {
        Arrays.sort(nums);
        List<Integer> dynamicList = new ArrayList<>();
        List<List<Integer>> answer = new ArrayList<>();
        
        go(answer, dynamicList, nums, 0, new StringBuilder());
        
        return answer;
    }
    void go(List<List<Integer>> answer, List<Integer> dynamicList, int[] nums, int index, StringBuilder sb) {
      
        set.add(dynamicList.toString());
        
        answer.add(new ArrayList<>(dynamicList));
        for (int i = index; i < nums.length; i++) {
            if (i != index && nums[i] == nums[i - 1])  continue;
            dynamicList.add(nums[i]);
            go(answer, dynamicList, nums, i + 1, sb);
            dynamicList.remove(dynamicList.size() - 1);
        }
    }
}

File: 0090-subsets-ii/0090-subsets-ii.go
func subsetsWithDup(nums []int) [][]int {
    var ans [][]int
    var curList []int
    bools := make([]bool, len(nums)+1)
    sort.Ints(nums)
    goo(&ans, curList, bools, nums, 0);
    
    return ans;
}

func goo(ans *[][]int, curList []int, bools []bool, nums []int, ind int) {
    
    *ans = append(*ans, append([]int{}, curList...))

    for i := ind; i < len(nums); i++ {
        if i != ind && nums[i] == nums[i - 1]  {
            continue; 
        }
        
        curList = append(curList, nums[i]);
        goo(ans, curList, bools, nums, i + 1);
        curList = curList[:len(curList)-1]
    }
}